cirq-core/cirq/sim/simulator.py

```diff
@@ -925,13 +926,10 @@ def __init__(
         self.params = params
         self.measurements = measurements
         self._final_step_result = final_step_result
-        self._final_simulator_state_cache = final_simulator_state
-
-    @property
-    def _final_simulator_state(self):
-        if self._final_simulator_state_cache is None:
-            self._final_simulator_state_cache = self._final_step_result._simulator_state()
-        return self._final_simulator_state_cache
+        self._final_simulator_state = (
```

Comment:
Why is this moved to the constructor? IIUC calling `_simulator_state()` on the final step result is a potentially-costly operation that was previously avoided if `_final_simulator_state` was never requested.

Reply:
Nice memory! Yes, previously `_simulator_state()` created e.g. `DensityMatrixSimulatorState`, which required merging. Now `_simulator_state()` just returns the reference to the `OperationTarget`, so it's safe. (Which brings up the point, there's really no reason to pass in `final_step_result` here anymore: we can always pass in the sim state directly. I'll deprecate that in a subsequent PR).